### PR TITLE
Restructure/transport solver

### PR DESCRIPTION
**Pull request by wkerzendorf**
This is merged next.

### CODE REVIEW

File: tardis/montecarlo/montecarlo_numba/base.py [hunk 1]
```diff
@@ -178,22 +160,15 @@ def montecarlo_main_loop(
...
-        np.ones_like(packet_collection.packets_output_nu, dtype=np.int64) * -1
-    )
-    last_line_interaction_out_ids = (
-        np.ones_like(packet_collection.packets_output_nu, dtype=np.int64) * -1
-    )
-    last_line_interaction_shell_ids = (
-        np.ones_like(packet_collection.packets_output_nu, dtype=np.int64) * -1
-    )
+    last_interaction_in_nus = np.empty(no_of_packets, dtype=np.float64)
+    last_interaction_types = -np.ones(no_of_packets, dtype=np.int64)
```
Comment (andrewfullard) [inline tardis/montecarlo/montecarlo_numba/base.py:168]: Putting the `-` before the `np.ones()` call is a little difficult to read. I would prefer `-1 * np.ones()`

File: tardis/montecarlo/montecarlo_numba/estimators.py [hunk 2]
```diff
@@ -65,24 +177,24 @@ def update_bound_free_estimators(
         photo_ion_rate_estimator_increment = (
             comov_energy * distance * x_sect_bfs[i] / comov_nu
         )
-        numba_estimator.photo_ion_estimator[
+        estimator_state.photo_ion_estimator[
```
Comment (andrewfullard) [inline tardis/montecarlo/montecarlo_numba/estimators.py:180]: Should this section be using the increment method instead?

File: tardis/montecarlo/montecarlo_numba/tests/test_base.py [hunk 3]
```diff
@@ -80,15 +74,14 @@ def test_montecarlo_main_loop(
...
     )
     npt.assert_allclose(actual_j_estimator, expected_j_estimator, rtol=1e-13)
-    npt.assert_allclose(actual_energy.value, expected_energy, rtol=1e-13)
-    npt.assert_allclose(actual_nu.value, expected_nu, rtol=1e-13)
+    npt.assert_allclose(actual_energy, expected_energy, rtol=1e-13)
+    npt.assert_allclose(actual_nu, expected_nu, rtol=1e-13)
 
 
-@pytest.mark.xfail(reason="need to store virtual packet data in hdf5")
+@pytest.mark.xfail
```
Comment (andrewfullard) [inline tardis/montecarlo/montecarlo_numba/tests/test_base.py:81]: Is this test still xfail? Why?

File: tardis/montecarlo/montecarlo_numba/tests/test_cuda_formal_integral.py [hunk 4]
```diff
@@ -21,7 +21,7 @@
     NumbaFormalIntegrator,
 )
 
-from tardis.montecarlo.base import MontecarloTransport
+from tardis.montecarlo.base import MontecarloTransportSolver
```
Comment (andrewfullard) [inline tardis/montecarlo/montecarlo_numba/tests/test_cuda_formal_integral.py:24]: Is this used here? I don't see the class name updated in the file

File: tardis/simulation/tests/test_simulation.py [hunk 5]
```diff
@@ -88,10 +88,26 @@ def simulation_one_loop(
...
-        actual = getattr(simulation_one_loop.transport, name)
-    except AttributeError:
+    if name in ["nu_bar_estimator", "j_estimator"]:
+        actual = getattr(
+            simulation_one_loop.transport.transport_state.estimators, name
+        )
+    elif name in ["t_radiative", "dilution_factor"]:
         actual = getattr(simulation_one_loop.simulation_state, name)
+    elif name in ["output_nu", "output_energy"]:
+        OLD_TO_NEW_DICT = {
```
Comment (andrewfullard) [inline tardis/simulation/tests/test_simulation.py:98]: Is this to avoid updating current refdata?

File: tardis/tests/fixtures/regression_data.py [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+
+    @property
+    def absolute_regression_data_dir(self):
+        return self.regression_data_path / self.relative_regression_data_dir
+
+    @property
+    def fpath(self):
+        return self.absolute_regression_data_dir / self.fname
+
+    def sync_dataframe(self, data, key="data"):
```
Comment (andrewfullard) [inline tardis/tests/fixtures/regression_data.py:62]: docstrings?

File: tardis/montecarlo/montecarlo_numba/tests/test_base.py (hunk 3 above)
Comment (wkerzendorf) [reply tardis/montecarlo/montecarlo_numba/tests/test_base.py:81]: I added this tests. it will no longer be xfail with the syrupy update

File: tardis/montecarlo/montecarlo_numba/base.py [hunk 7]
```diff
@@ -178,22 +160,15 @@ def montecarlo_main_loop(
     virtual_packet_logging : bool
         Option to enable virtual packet logging.
     """
-    output_nus = np.empty_like(packet_collection.packets_input_nu)
-    last_interaction_types = (
-        np.ones_like(packet_collection.packets_output_nu, dtype=np.int64) * -1
-    )
-    output_energies = np.empty_like(packet_collection.packets_output_nu)
+    no_of_packets = len(packet_collection.initial_nus)
+    output_nus = np.empty(no_of_packets, dtype=np.float64)
```
Comment (Knights-Templars) [inline tardis/montecarlo/montecarlo_numba/base.py:164]: Why are you initializing with np.empty? Will np.zeros also work

File: tardis/simulation/tests/test_simulation.py (hunk 5 above)
Comment (wkerzendorf) [reply tardis/simulation/tests/test_simulation.py:98]: yes

File: tardis/tests/fixtures/regression_data.py (hunk 6 above)
Comment (wkerzendorf) [reply tardis/tests/fixtures/regression_data.py:62]: added now

File: tardis/montecarlo/montecarlo_numba/base.py (hunk 1 above)
Comment (wkerzendorf) [reply tardis/montecarlo/montecarlo_numba/base.py:168]: done

File: tardis/montecarlo/base.py [hunk 8]
```diff
@@ -113,10 +90,10 @@ def __init__(
         self.v_packet_settings = v_packet_settings
         self.spectrum_method = spectrum_method
         self._integrator = None
-        self._spectrum_integrated = None
+
         self.use_gpu = use_gpu
 
-        self.virt_logging = virtual_packet_logging
+        self.virt_logging = enable_virtual_packet_logging
         self.virt_packet_last_interaction_type = np.ones(2) * -1
```
Comment (wkerzendorf) [inline tardis/montecarlo/base.py:97]: comment why length 2

File: tardis/montecarlo/base.py [hunk 9]
```diff
@@ -323,47 +172,51 @@ def run(
...
+        self._initialize_packets(no_of_packets, iteration)
 
-        self._initialize_packets(
-            simulation_state,
-            no_of_packets,
-            iteration,
+        self.transport_state = MonteCarloTransportState(
+            self.packet_collection,
+            estimators,
+            simulation_state.volume.cgs.copy(),
```
Comment (wkerzendorf) [inline tardis/montecarlo/base.py:190]: geometry state and volume duplicate - fix I think in the next one.

File: tardis/montecarlo/montecarlo_numba/estimators.py (hunk 2 above)
Comment (wkerzendorf) [reply tardis/montecarlo/montecarlo_numba/estimators.py:180]: no, the increment method is just adding two estimator classes from two different threads. However, a much more detailed restructure of this is done in #2500 where we will document it.

File: tardis/montecarlo/montecarlo_numba/tests/test_cuda_formal_integral.py (hunk 4 above)
Comment (wkerzendorf) [reply tardis/montecarlo/montecarlo_numba/tests/test_cuda_formal_integral.py:24]: ran ruff. This is skipped on non-cuda machines and we should track that.

File: tardis/montecarlo/base.py (hunk 8 above)
Comment (wkerzendorf) [reply tardis/montecarlo/base.py:97]: done

File: tardis/montecarlo/montecarlo_numba/tests/test_base.py (hunk 3 above)
Comment (andrewfullard) [reply tardis/montecarlo/montecarlo_numba/tests/test_base.py]: Can this xfail be removed then to make sure it passes?

Comment (wkerzendorf) [reply tardis/montecarlo/montecarlo_numba/tests/test_base.py]: this is no longer xfail - it's outdated

File: tardis/montecarlo/base.py (hunk 9 above)
Comment (wkerzendorf) [reply tardis/montecarlo/base.py:192]: it is fixed in the next one

**Review by andrewfullard — DISMISSED** [review]

File: tardis/grid/tests/test_grid.py [hunk 10]
```diff
@@ -4,8 +4,11 @@
 import pandas as pd
 
 import tardis
+
+from pathlib import Path
```
Comment (jvshields) [inline tardis/grid/tests/test_grid.py:8]: Looks like this is being double imported - check line 1

File: tardis/io/model_reader.py
Comment (jvshields) [inline tardis/io/model_reader.py:1]: On line 675 - I think part of this restructure is trying to make sure we use more informative names. transport_from_hdf mostly operates on a dictionary called d, but it might be worth changing this to a more descriptive name. Not sure if this is outside the scope of this restructure though.

File: tardis/model/parse_input.py [hunk 11]
```diff
@@ -5,6 +5,7 @@
 import pandas as pd
 from astropy import units as u
 
+from tardis import constants as const
```
Comment (jvshields) [inline tardis/model/parse_input.py:8]: Double import here. Also imported immediately below in line 9

File: tardis/montecarlo/base.py [hunk 12]
```diff
@@ -31,29 +28,19 @@
 
 
 # TODO: refactor this into more parts
-class MontecarloTransport(HDFWriterMixin):
+class MontecarloTransportSolver(HDFWriterMixin):
```
Comment (jvshields) [inline tardis/montecarlo/base.py:31]: Should the C be capitalized for correct camel case?

**Review by andrewfullard — APPROVED** [review]
I'll pat myself on the back here
